### PR TITLE
Add `tp_new` slot to `VariableTracker` and support it in user-inherited containers.

### PR DESCRIPTION
**Pull request by hameerabbasi**
Side-comment: The `functools.lru_cache` and `io.StringIO` tests need more work (`VariableTrackers` modelling the respective types).

Towards #195901

cc @voznesenskym @penguinwu @EikanWang @jgong5 @Guobing-Chen @XiaobingSuper @zhuhaozhe @blzheng @wenzhe-nrv @jiayisunx @kadeng @chauhang @amjames @jataylo @azahed98

### CODE REVIEW

File: torch/_dynamo/variables/builtin.py [hunk 1]
```diff
@@ -1891,7 +1891,40 @@ def call_method(
...
+                # frozenset.__init__ is a no-op. Unlike set above, the
+                # iterable must be kept, mirroring tuple's handling above.
+                if is_exact_type:
+                    init_args = unpack_iterable(tx, args[1]) if len(args) > 1 else []
+                    return FrozensetVariable(
+                        init_args, mutation_type=ValueMutationNew()
+                    )
+                return tx.output.side_effects.track_new_user_defined_object(
+                    self, args[0], args[1:], tx=tx
+                )
```
Comment (guilhermeleobas) [inline torch/_dynamo/variables/builtin.py]: What do you think of introducing [tp_new](https://docs.python.org/3/c-api/typeobj.html#c.PyTypeObject.tp_new) to VariableTracker to model this? Then, this becomes:

```python
if name == "__new__":
    return args[0].tp_new(...)

File: torch/_dynamo/variables/builtin.py [hunk 2]
```diff
@@ -1816,6 +1816,61 @@ def call_function(
...
+        self,
+        tx: "InstructionTranslatorBase",
+        args: list[VariableTracker],
+        kwargs: dict[str, VariableTracker],
+    ) -> VariableTracker:
+        if self.fn is object and len(args) == 1:
+            if len(kwargs) != 0:
+                raise AssertionError(
+                    f"object.__new__ expects no kwargs, got {len(kwargs)}"
+                )
```
Comment (guilhermeleobas) [inline torch/_dynamo/variables/builtin.py:1829]: Use `no_keywords` here instead.

File: torch/_dynamo/variables/builtin.py [hunk 3]
```diff
@@ -1816,6 +1816,61 @@ def call_function(
...
+            return tx.output.side_effects.track_new_user_defined_object(
+                self, args[0], args[1:], tx=tx
+            )
```
Comment (guilhermeleobas) [inline torch/_dynamo/variables/builtin.py:1832]: I have the impression that this should be on `VariableTracker`

File: torch/_dynamo/variables/builtin.py [hunk 4]
```diff
@@ -1816,6 +1816,61 @@ def call_function(
...
+
+        if self.fn is tuple and len(args) == 2 and not kwargs:
+            if isinstance(args[0], BuiltinVariable) and args[0].fn is tuple:
+                init_args = unpack_iterable(tx, args[1])
+                return variables.TupleVariable(
+                    init_args, mutation_type=ValueMutationNew()
+                )
+            return tx.output.side_effects.track_new_user_defined_object(
+                self, args[0], args[1:], tx=tx
+            )
```
Comment (guilhermeleobas) [inline torch/_dynamo/variables/builtin.py:1842]: Same thing for the rest of this function. Perhaps this should be on TupleVariable

File: torch/_dynamo/variables/base.py [hunk 5]
```diff
@@ -2336,6 +2336,28 @@ def tp_init_impl(
             hints=[*graph_break_hints.DYNAMO_BUG],
         )
 
+    def tp_new_impl(
```
Comment (guilhermeleobas) [inline torch/_dynamo/variables/base.py:2339]: You should also uncomment the `# SlotDef("__new__", ...)` entry on slotdefs

File: torch/_dynamo/variables/user_defined.py [hunk 6]
```diff
@@ -1082,6 +1082,37 @@ def fake_cross_entropy_loss(
 
         return variables.LambdaVariable(fake_cross_entropy_loss)
 
+    def tp_new_impl(
+        self,
+        tx: "InstructionTranslatorBase",
+        args: list[VariableTracker],
+        kwargs: dict[str, VariableTracker],
+    ) -> VariableTracker:
+        # Caller must have already checked
```
Comment (guilhermeleobas) [inline torch/_dynamo/variables/user_defined.py:1091]: Here we need to model [`slot_tp_new`](https://github.com/python/cpython/blob/52ffffe0a23bf0f4a57ee00377c5aeb965b3a29a/Objects/typeobject.c#L11129-L11147) here.

I think the rules for `__new__` here should be:
1) Call `__new__` if the UDOV declares it
2) If `_base_vt` is defined, call `_base_vt.__new__`.
3) If none are defined, `object.__new__` should be used.

**Review by guilhermeleobas — COMMENTED** [review]
Great work. Perhaps we should look at the CPython tests for `__new__` to cover any corner case as well.

File: torch/_dynamo/variables/base.py [hunk 7]
```diff
@@ -2345,6 +2360,35 @@ def tp_init_impl(
...
+        implements `object.__new__(cls)`, the ultimate fallback every type's
+        `__new__` chain bottoms out at.
+        """
+        if len(args) != 1:
+            unimplemented(
+                gb_type="missing tp_new",
+                context=f"tp_new_impl not implemented for {self.python_type_name()}",
+                explanation=f"Dynamo does not know how to trace __new__ on `{self.debug_repr()}`.",
+                hints=[*graph_break_hints.DYNAMO_BUG],
+            )
```
Comment (guilhermeleobas) [inline torch/_dynamo/variables/base.py]: Raise TypeError instead?

```c
static PyObject *
object_new(PyTypeObject *type, PyObject *args, PyObject *kwds)
{
    if (excess_args(args, kwds)) {
        if (type->tp_new != object_new) {
            PyErr_SetString(PyExc_TypeError,
                            "object.__new__() takes exactly one argument (the type to instantiate)");
            return NULL;
        }
        if (type->tp_init == object_init) {
            PyErr_Format(PyExc_TypeError, "%.200s() takes no arguments",
                         type->tp_name);
            return NULL;
        }
    }
    ...
```

File: torch/_dynamo/variables/user_defined.py [hunk 8]
```diff
@@ -1143,6 +1143,84 @@ def fake_cross_entropy_loss(
...
+                ),
+                hints=[*graph_break_hints.SUPPORTABLE],
+            )
+        if self.value is collections.OrderedDict:
+            # Exact OrderedDict: represent as a bare OrderedDictVariable,
+            # mirroring dict.__new__(dict) -> ConstDictVariable. OrderedDict
+            # does not declare its own __new__ (it inherits dict.__new__),
+            # so this must be special-cased ahead of the MRO walk below --
+            # there is no dedicated OrderedDictBuiltinVariable to delegate to.
+            return OrderedDictVariable({}, mutation_type=ValueMutationNew())
```
Comment (guilhermeleobas) [inline torch/_dynamo/variables/user_defined.py]: Does a OrderedDict go through this path?

File: torch/_dynamo/variables/user_defined.py
Comment (guilhermeleobas) [inline torch/_dynamo/variables/user_defined.py:1]: This file is missing a `tp_new_impl` for UserDefinedObjectVariable I think. It should mimic the `slot_tp_new`  function:

```c
static PyObject *
slot_tp_new(PyTypeObject *type, PyObject *args, PyObject *kwds)
{
    PyThreadState *tstate = _PyThreadState_GET();
    PyObject *func, *result;

    func = PyObject_GetAttr((PyObject *)type, &_Py_ID(__new__));
    if (func == NULL) {
        return NULL;
    }

    result = _PyObject_Call_Prepend(tstate, func, (PyObject *)type, args, kwds);
    Py_DECREF(func);
    return result;
}
```

File: torch/_dynamo/variables/base.py (hunk 7 above)
Comment (hameerabbasi) [reply torch/_dynamo/variables/base.py]: Done in https://github.com/pytorch/pytorch/pull/196425/commits/42a81f77672e4a85f44311760d66da0eeaf4fb09

File: torch/_dynamo/variables/user_defined.py
Comment (hameerabbasi) [reply torch/_dynamo/variables/user_defined.py:1]: Done in https://github.com/pytorch/pytorch/pull/196425/commits/42a81f77672e4a85f44311760d66da0eeaf4fb09.

File: torch/_dynamo/variables/user_defined.py (hunk 8 above)
Comment (hameerabbasi) [reply torch/_dynamo/variables/user_defined.py]: It does, but it turns out to be completely unnecessary. Resolved in https://github.com/pytorch/pytorch/pull/196425/commits/cfa974d651c56127565f0b82d6e49d5bbc023af7.

File: torch/_dynamo/variables/user_defined.py
Comment (hameerabbasi) [reply torch/_dynamo/variables/user_defined.py:1]: Actually, I'll remove this; since `tp_new_impl` can only be defined on classes, not on objects. I ran a full CPython test suite run with a breakpoint in here; it wasn't ever hit.

Edit: Removed in https://github.com/pytorch/pytorch/pull/196425/commits/f652ae24dd300c5f351461f29b745605551297f5.

**Comment by guilhermeleobas** [comment]
@hameerabbasi claude found one regression in this PR:

```python
class MySet(set):
    def __new__(cls, it):
        return super().__new__(cls, it)

def f():
    s = MySet([1, 2])
    torch._dynamo.graph_break()
    return sorted(s)

f()                                   # [1, 2]
torch.compile(f, backend="eager")()   # this PR: []   before this PR: [1, 2]
```

And a minor bug in tuple:
> `tuple.__new__(T)` (1 arg, legal, returns ()) graph-breaks. Fix mirrors FrozensetVariable's len(args) > 1
